tool_path_planner/src/edge_path_generator.cpp

```diff
@@ -0,0 +1,589 @@
+{
+  setInput(mesh);
+  return generate(config);
+}
+
+boost::optional<std::vector<geometry_msgs::PoseArray>>
+EdgePathGenerator::generate(pcl::PolygonMesh::ConstPtr mesh, const tool_path_planner::EdgePathConfig& config)
+{
+  setInput(mesh);
+  return generate(config);
+}
+
+int EdgePathGenerator::mergePoints(const tool_path_planner::EdgePathConfig& config,const pcl::PointIndices& edge_segment,
+                                    pcl::PointCloud<pcl::PointNormal>& merged_points)
+{
+  using namespace Eigen;
+  using namespace pcl;
+  using PType = std::remove_reference<decltype(merged_points)>::type::PointType;
+  decltype(edge_segment.indices) merged_indices;
+  Vector3f dir;
+  PType mid_point;
+
+
+  auto compute_dist = [&](const PType& p1, const PType& p2) -> double
+  {
+    dir = p2.getVector3fMap() - p1.getVector3fMap();
+    return dir.norm();
+  };
+
+
+  for(std::size_t i = 0; i < edge_segment.indices.size()-1; i++)
+  {
+    int idx_current = edge_segment.indices[i];
+    int idx_next = edge_segment.indices[i+1];
+
+    if(std::find(merged_indices.begin(), merged_indices.end(), idx_current) != merged_indices.end() )
+    {
+      // already merged so continue
+      continue;
+    }
+
+    const PType& p1 = (*input_cloud_)[idx_current];
+    const PType& p2 = (*input_cloud_)[idx_next];
+    dir = p2.getVector3fMap() - p1.getVector3fMap();
+    if(compute_dist(p1, p2) >  config.merge_dist)
+    {
+      // adding current point and continue
+      merged_points.push_back(p1);
+      continue;
+    }
+
+    mid_point.getVector3fMap() = p1.getVector3fMap() + 0.5 * dir;
+
+    mid_point.getNormalVector3fMap() = (p2.getNormalVector3fMap() + p1.getNormalVector3fMap()).normalized();
+    merged_points.push_back(mid_point);
+    merged_indices.push_back(idx_next);
+  }
+
+  // checking first and last point
+  const PType& p0 = (*input_cloud_)[edge_segment.indices.front()];
+  const PType& pf = (*input_cloud_)[edge_segment.indices.back()];
+  if(compute_dist(p0, merged_points.front()) > 1e-6)
+  {
+    merged_points.insert(merged_points.begin(),p0 );
+  }
+
+  if(compute_dist(pf, merged_points.back()) > 1e-6)
+  {
+    merged_points.push_back(pf);
+  }
+
+  return merged_indices.size();
+}
+
+bool EdgePathGenerator::splitEdgeSegments(const tool_path_planner::EdgePathConfig& config,
+                                          const pcl::PointIndices& edge_indices,
+                                          std::vector<pcl::PointIndices>& segments)
+{
+  using namespace Eigen;
+  using namespace pcl;
+  using AlignedPointTVector = std::remove_reference<decltype(*this->octree_search_)>::type::AlignedPointTVector;
+  using PType = std::remove_reference<decltype(*this->input_cloud_)>::type::PointType;
+  struct Ray
+  {
+    PType p;
+    Vector3f d;
+  };
+
+  // checks for surfaces between the points and returns > 0 (num of voxels_encountered) when there's one
+  auto check_surface_intersection = [&](const PType& p1, const PType& p2) -> int
```

Comment:
I recommend moving this out into a separate function.

Reply:
I will do that

Reply:
Actually I think this is fine. See my comment on PR #70